File: fixtures/scale/dart/large/lib/ciphers/keywordsub/keywordsub_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'keywordsub_key.dart';

/// Monoalphabetic substitution built from a keyword-mixed alphabet.
class KeywordSubCipher implements Cipher {
  final String _mixed;

  KeywordSubCipher(KeywordSubKey key) : _mixed = key.mixedAlphabet();

  @override
  String name() => 'keywordsub';

  @override
  String encrypt(String plaintext) {
    final sb = StringBuffer();
    for (final c in cleanAlphabet(plaintext).split('')) {
      sb.write(_mixed[indexOf(c)]);
    }
    return sb.toString();
  }

  @override
  String decrypt(String ciphertext) {
    final sb = StringBuffer();
    for (final c in cleanAlphabet(ciphertext).split('')) {
      sb.write(charAt(_mixed.indexOf(c)));
    }
    return sb.toString();
  }
}

File: fixtures/scale/scala/large/src/cryptobench/ciphers/keywordsub/KeywordSubCipher.scala
package cryptobench.ciphers.keywordsub

import cryptobench.core.Cipher
import cryptobench.util.Alphabet

/** Monoalphabetic substitution built from a keyword-mixed alphabet. */
final class KeywordSubCipher(key: KeywordSubKey) extends Cipher {
  private val mixed = key.mixedAlphabet

  override def name: String = "keywordsub"

  override def encrypt(plaintext: String): String = {
    val sb = new StringBuilder
    for (c <- Alphabet.clean(plaintext)) {
      sb.append(mixed.charAt(Alphabet.indexOf(c)))
    }
    sb.toString
  }

  override def decrypt(ciphertext: String): String = {
    val sb = new StringBuilder
    for (c <- Alphabet.clean(ciphertext)) {
      sb.append(Alphabet.charAt(mixed.indexOf(c)))
    }
    sb.toString
  }
}
